homeassistant/helpers/update_coordinator.py

```diff
@@ -30,12 +31,14 @@
 REQUEST_REFRESH_DEFAULT_COOLDOWN = 10
 REQUEST_REFRESH_DEFAULT_IMMEDIATE = True
 
-_T = TypeVar("_T")
+_DataT = TypeVar("_DataT", default=dict[str, Any])
```

Comment:
Will this type the coordinator data as `dict[str, Any]` even if the implementing integration haven't added type annotations? That would be misleading and incorrect I think.

Reply:
> Will this type the coordinator data as `dict[str, Any]` even if the implementing integration hasn't added type annotations? That would be misleading and incorrect I think.

Yes, but that's by design. A lot of times, `CoordinatorEntity` is used without any generic type. That causes mypy to skip all subsequent checks for `self.coordinator.data` as the type is assumed to be `Any`. In reality most integrations start out with `dict[str, Any]`. However, once the integration grows, the data type is often changed to a custom one like TypedDict or DataClass. Even then, the set `default` will be helpful. Once mypy fully supports PEP 696, it will warn that for example `data` doesn't have a specific attribute. There will also be a message that the type for `_async_update_data` doesn't match `dict[str, Any]`. Both of these can be used as hints to update the generic type for `CoordinatorEntity` / `DataUpdateCoordinator`.

#### Some examples

`Airthings` still uses a `dict[str, Any]`.
https://github.com/home-assistant/core/blob/8987e023a0b7d93b167859ee40758ea3e15430b3/homeassistant/components/airthings/sensor.py#L133
https://github.com/home-assistant/core/blob/8987e023a0b7d93b167859ee40758ea3e15430b3/homeassistant/components/airthings/sensor.py#L163-L166

I mentioned the PR for `notion` earlier already #91883. Say we remove all generic arguments for `CoordinatorEntity` and `DataUpdateCoordinator`, with a future mypy version we would get these errors:
```bash
homeassistant/components/notion/__init__.py:290: error: "dict[str, Any]" has no attribute "bridges"  [attr-defined]
homeassistant/components/notion/__init__.py:291: error: "dict[str, Any]" has no attribute "sensors"  [attr-defined]
homeassistant/components/notion/__init__.py:314: error: "dict[str, Any]" has no attribute "listeners"  [attr-defined]
homeassistant/components/notion/__init__.py:320: error: Returning Any from function declared to return "Listener"  [no-any-return]
homeassistant/components/notion/__init__.py:320: error: "dict[str, Any]" has no attribute "listeners"  [attr-defined]
homeassistant/components/notion/__init__.py:328: error: "dict[str, Any]" has no attribute "sensors"  [attr-defined]
homeassistant/components/notion/__init__.py:335: error: "dict[str, Any]" has no attribute "bridges"  [attr-defined]
homeassistant/components/notion/__init__.py:343: error: "dict[str, Any]" has no attribute "bridges"  [attr-defined]
homeassistant/components/notion/__init__.py:356: error: "dict[str, Any]" has no attribute "listeners"  [attr-defined]
homeassistant/components/notion/sensor.py:70: error: "dict[str, Any]" has no attribute "user_preferences"  [attr-defined]
homeassistant/components/notion/sensor.py:72: error: "dict[str, Any]" has no attribute "user_preferences"  [attr-defined]
homeassistant/components/notion/diagnostics.py:43: error: "dict[str, Any]" has no attribute "asdict"  [attr-defined]
```

Whereas without we only get one and wouldn't even know that the generic type is wrong
```bash
homeassistant/components/notion/__init__.py:320: error: Returning Any from function declared to return "Listener"  [no-any-return]
```

--
Over the last few weeks, I've already added generic arguments to integrations which use (strict) type checking and have a custom data / DataUpdateCoordinator type. At least the issues I could find are fixed.

Reply:
If we don't get any warnings of incorrect type with the current mypy version the default may just silently be wrong and misleading. I don't think we should add it until we get warnings.

Reply:
> If we don't get any warnings of incorrect type with the current mypy version the default may just silently be wrong and misleading. I don't think we should add it until we get warnings.

I'm working on the mypy support, hopefully the warning will be there with the release next month 🤞🏻
In the meantime, it's still an improvement as I suspect a lot / most use VS Code and Pylance which will already display the errors (if they have enabled the type checking mode).

I'm not planning to add any more defaults until then, just for these once I think the benefits outweigh the small risk of it becoming outdated.

Reply:
I disagree that the benefits outweigh the problems and think we should remove this for now.

Reply:
Guess we have different opinions there. Anyway you call the shots. Opened #95101